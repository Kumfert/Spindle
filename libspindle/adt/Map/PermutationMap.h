//
//  PermutationMap.h -- an abstraction of a permutation vector.
//
//  $Id: PermutationMap.h,v 1.2 2000/02/18 01:31:49 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  The PermutationMap class is meant to abstract away the concept of
//  a permutation vector which may be of old2New or new2old varieties.
//
//  A PermutationMap is not in a valid state until validate() is called
//  to insure that both old2New and new2Old are consistent with each
//  other.
//

#ifndef SPINDLE_PERMUTATION_H_
#define SPINDLE_PERMUTATION_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class PermutationMap : public SpindlePersistant {
  SPINDLE_DECLARE_PERSISTANT( PermutationMap )

private:
  PermutationMap();
  SharedArray<int> old2New;
  SharedArray<int> new2Old;

  enum { NEITHER, OLD2NEW, NEW2OLD } changedMostRecent;
  int sz;
public:
  // constructors/destructors
  PermutationMap( const int n );
  PermutationMap( const PermutationMap& perm );
  virtual ~PermutationMap();

  int size() const;

  // access specific permuation vectors
  const SharedArray<int>& getOld2New() const;
  SharedArray<int>& getOld2New();

  const SharedArray<int>& getNew2Old() const;
  SharedArray<int>& getNew2Old();


  virtual void validate();
  virtual void dump( FILE * stream ) const;

  virtual bool reset();
  void resize( const int n );
  void prettyPrint( FILE * fp, bool new2old_perm=true, bool printLength=false, 
		    bool onesOffset=false ) const; 

};

inline 
int PermutationMap::size() const {
  return sz;
}

inline 
PermutationMap::PermutationMap( const int n )
  : old2New(n), new2Old(n), sz(n) {
    incrementInstanceCount( PermutationMap::MetaData );
    reset();
}

inline 
PermutationMap::PermutationMap( const PermutationMap& perm ) 
  : sz( perm.size() ) {
    incrementInstanceCount( PermutationMap::MetaData );
    old2New.import( perm.old2New );
    new2Old.import( perm.new2Old );
    changedMostRecent = perm.changedMostRecent;
    currentState = perm.currentState;
}

inline 
PermutationMap::~PermutationMap() {
  decrementInstanceCount( PermutationMap::MetaData );
}

SPINDLE_END_NAMESPACE

#endif


