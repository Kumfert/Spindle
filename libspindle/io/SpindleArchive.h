//
//  SpindleArchive.h
//
// $Id: SpindleArchive.h,v 1.2 2000/02/18 01:31:51 kumfert Exp $
//
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  This class is used by all persistant objects to store their data
//

#ifndef SPINDLE_ARCHIVE_H_
#define SPINDLE_ARCHIVE_H_


#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifdef REQUIRE_OLD_CXX_HEADER_SUFFIX
#include "pair.h"
#else
#include <utility>
using std::pair;
#endif

// This is a strictly C++ class 
#include <string>
using std::string;

SPINDLE_BEGIN_NAMESPACE

struct ClassMetaData;
class SpindleBaseClass;
class SpindleFile;

/**
 * @memo Used by all persistant objects to store/load their data
 * @type class
 *
 * @author Gary Kumfert
 * @version  #$Id: SpindleArchive.h,v 1.2 2000/02/18 01:31:51 kumfert Exp $#
 * @see SpindleFile
 * @see SpindlePersistant
 */
class SpindleArchive { 
public:
  /// different modes for using an archive object
  enum Mode { store=0x01, 
	      load =0x02,
	      noFlushOnDelete =0x04,
	      noByteSwap = 0x08
            };
  /// constructor
  SpindleArchive( SpindleFile* fp, unsigned int mode, int buffSize=4096, void* buffer=NULL );
  /// destructor
  ~SpindleArchive();

  /**@name simple tests */
  //@{
  /// returns true iff the archive is for loading new instances.
  bool isLoading() const;
  /// returns true iff the archive is storing existing instances.
  bool isStoring() const;
  /// returns true iff the archive is swapping bytes for cross-platform compatability
  bool isByteSwapping() const;
  /// returns true iff the buffer is empty
  bool isBufferEmpty() const;
  //@}

  /// returns the SpindleFile instance that the SpindleArchive is using.
  SpindleFile* getFile();

  void fillBuffer( int atLeast=-1 );
  void flush();
  void close();

  size_t peekLength();
  size_t read( void* buf, size_t itemSz, size_t nItems );
  void write(const void* buf, size_t itemSz, size_t nItems );

  const ClassMetaData* readObjectID();
  SpindleBaseClass* readObject( const ClassMetaData* metaData );
  void writeObject( const SpindleBaseClass* object );

  /**@name insertion */
  //@{
  /**@name friend insertion operators */
  //@{
  /// stores a zero terminated character array
  friend SpindleArchive& operator<<( SpindleArchive& ar, const char* str );
  /// stores a C++ string
  friend SpindleArchive& operator<<( SpindleArchive& ar, const string& str );
  /// stores a class derived from SpindleBaseClass
  friend SpindleArchive& operator<<( SpindleArchive& ar, const SpindleBaseClass* object );
  //@}
  /**@name single value insertion*/
  //@{
  /// stores a char
  SpindleArchive& operator<<( const signed char t);
  /// stores an unsigned char
  SpindleArchive& operator<<( const unsigned char t);
  /// stores a short
  SpindleArchive& operator<<( const signed short t);
  /// stores an unsigned short
  SpindleArchive& operator<<( const unsigned short t);
  /// stores an int
  SpindleArchive& operator<<( const signed int t);
  /// stores an unsigned int
  SpindleArchive& operator<<( const unsigned int t);
  /// stores a long
  SpindleArchive& operator<<( const signed long t);
  /// stores an unsigned long
  SpindleArchive& operator<<( const unsigned long t);
  /// stores a float
  SpindleArchive& operator<<( const float t);
  /// stores a double
  SpindleArchive& operator<<( const double t);
  /// stores a long double
  SpindleArchive& operator<<( const long double t);
  //@}
  /**@name array insertion */
  //@{
  /// stores and array of char's
  SpindleArchive& operator<<( const pair<const signed char*, size_t>& p);
  /// stores and array of unsigned char's
  SpindleArchive& operator<<( const pair<const unsigned char*, size_t>& p);
  /// stores and array of short's
  SpindleArchive& operator<<( const pair<const signed short*, size_t>& p);
  /// stores and array of unsigned short's
  SpindleArchive& operator<<( const pair<const unsigned short*, size_t>& p);
  /// stores and array of int's
  SpindleArchive& operator<<( const pair<const signed int*, size_t>& p);
  /// stores and array of unsigned int's
  SpindleArchive& operator<<( const pair<const unsigned int*, size_t>& p);
  /// stores and array of long's
  SpindleArchive& operator<<( const pair<const signed long*, size_t>& p);
  /// stores and array of unsigned long's
  SpindleArchive& operator<<( const pair<const unsigned long*, size_t>& p);
  /// stores and array of floats
  SpindleArchive& operator<<( const pair<const float*, size_t>& p);
  /// stores and array of doubles
  SpindleArchive& operator<<( const pair<const double*, size_t>& p);
  /// stores and array of long doubles
  SpindleArchive& operator<<( const pair<const long double*, size_t>& p);
  //@}
  //@}
  
  /**@name extraction */
  //@{
  /**@name friend extraction operators */
  //@{
  /// loads a zero terminated character string
  friend SpindleArchive& operator>>( SpindleArchive& ar, char* &str );
  /// loads a C++ string
  friend SpindleArchive& operator>>( SpindleArchive& ar, string& str );
  /// loads an instance of a class derived from SpindleBaseClass
  friend SpindleArchive& operator>>( SpindleArchive& ar, SpindleBaseClass* &object );
  //@}
  /**@name single value extraction */
  //@{
  /// loads a char
  SpindleArchive& operator>>( signed char& );
  /// loads an unsigned char
  SpindleArchive& operator>>( unsigned char& );
  /// loads a short
  SpindleArchive& operator>>( signed short& );
  /// loads an unsigned short
  SpindleArchive& operator>>( unsigned short& );
  /// loads an int
  SpindleArchive& operator>>( signed int& );
  /// loads an unsigned int
  SpindleArchive& operator>>( unsigned int& );
  /// loads a long
  SpindleArchive& operator>>( signed long& );
  /// loads an unsigned long
  SpindleArchive& operator>>( unsigned long& );
  /// loads a float
  SpindleArchive& operator>>( float& );
  /// loads a double
  SpindleArchive& operator>>( double& );
  /// loads a long double
  SpindleArchive& operator>>( long double& );
  //@}
  /**@name  array extraction */
  //@{
  /// loads an array of char's
  SpindleArchive& operator>>( pair< signed char*, size_t>& );
  /// loads an array of unsigned char's
  SpindleArchive& operator>>( pair< unsigned char*, size_t>& );
  /// loads an array of short's
  SpindleArchive& operator>>( pair< signed short*, size_t>& );
  /// loads an array of unsigned short's
  SpindleArchive& operator>>( pair< unsigned short*, size_t>& );
  /// loads an array of int's
  SpindleArchive& operator>>( pair< signed int*, size_t>& );
  /// loads an array of unsigned int's
  SpindleArchive& operator>>( pair< unsigned int*, size_t>& );
  /// loads an array of long's
  SpindleArchive& operator>>( pair< signed long*, size_t>& );
  /// loads an array of unsigned long's
  SpindleArchive& operator>>( pair< unsigned long*, size_t>& );
  /// loads an array of float's
  SpindleArchive& operator>>( pair< float*, size_t>& );
  /// loads an array of double's
  SpindleArchive& operator>>( pair< double*, size_t>& );
  /// loads an array of long double's
  SpindleArchive& operator>>( pair< long double*, size_t>& );
  //@}
  //@}
protected:
  /// copy constructor
  SpindleArchive( const SpindleArchive& src );
  /// assignment
  void operator=( const SpindleArchive& src );

  unsigned int mode_;
  size_t bufferSize_;
  SpindleFile* file_;
  
  char* bufCur_;
  char* bufMax_;
  char* bufStart_;
  bool ownBuffer_;

};

SPINDLE_END_NAMESPACE

#endif
