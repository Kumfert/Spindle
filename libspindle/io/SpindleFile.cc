//
// SpindleFile.cc -- raw unbuffered binary I/O file
//
// $Id: SpindleFile.cc,v 1.2 2000/02/18 01:31:51 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//

#include "spindle/spindle.h"

#include "spindle/SpindleFile.h"

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#include <string.h>

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

 
#ifdef COMPRESS
#define DO_COMPRESS    COMPRESS " - -c > %s"
#endif
#ifdef UNCOMPRESS
#define DO_UNCOMPRESS  UNCOMPRESS " -c %s"
#endif

#ifdef GZIP
#define DO_GZIP        GZIP   " -9 - -c > %s"
#endif
#ifdef GUNZIP
#define DO_GUNZIP      GUNZIP " -c %s"
#endif GUNZIP 


#ifdef PKZIP
#define DO_PKZIP       PKZIP   " %s -q -"
#endif 
#ifdef PKUNZIP
#define DO_PKUNZIP     PKUNZIP " -p %s"
#endif

#ifdef _MSC_VER

#define popen _popen
#define pclose _pclose

#endif

SPINDLE_IMPLEMENT_DYNAMIC( SpindleFile, SpindleBaseClass )

#ifdef __FUNC__
#undef __FUNC__
#endif

#define __FUNC__ "SpindleFile::SpindleFile()"
SpindleFile::SpindleFile() { 
  FENTER;
  incrementInstanceCount( SpindleFile::MetaData );
  closeOnDelete_ = dont;
  filename_ = 0;
  fp_ = 0;
  FEXIT;
}
#undef __FUNC__


#define __FUNC__ "SpindleFile::SpindleFile( FILE * fp )"
SpindleFile::SpindleFile( FILE * fp ) {
  FENTER;
  ASSERT(fp != 0, "FILE pointer cannot be null.");

  incrementInstanceCount( SpindleFile::MetaData );
  closeOnDelete_ = dont;
  filename_ = 0;
  fp_ = fp;
  FEXIT;
}
#undef __FUNC__

#define __FUNC__ "SpindleFile::SpindleFile( int filedes, const char * openFlags )"
SpindleFile::SpindleFile( int filedes, const char * openFlags ) {
  FENTER;
  ASSERT( filedes > 0, "Negative file descriptor.");

  incrementInstanceCount( SpindleFile::MetaData );
  if (! open( filedes, openFlags ) ) { 
    filename_ = 0;
    //    currentState_ = INVALID;
    FEXIT;
  }
  filename_ = "unknown";
  //  currentState_ = VALID;
  FEXIT;
}
#undef __FUNC__

#define __FUNC__ "SpindleFile::SpindleFile( const char* filename, const char* openFlags )"
SpindleFile::SpindleFile( const char* filename, const char* openFlags ) {
  FENTER;
  ASSERT( filename!=0, "`filename' is a null pointer.");
  ASSERT( openFlags!=0, "`openFlags' is a null pointer.");

  incrementInstanceCount( SpindleFile::MetaData );
  if ( ! this->open( filename, openFlags ) ) { 
    filename_ = 0;
    //    currentState_ = INVALID;
    FEXIT;
  }
  filename_  = filename;
  //  currentState_ = VALID;
  FEXIT;
}
#undef __FUNC__

#define __FUNC__ "bool SpindleFile::open( int filedes, const char* openFlags )"
bool SpindleFile::open( int filedes, const char* openFlags ) {
  FENTER;
  if ((fp_ = fdopen( filedes, openFlags )) == NULL ) {
    //    currentState_ = INVALID;
    closeOnDelete_ = dont;
    FRETURN(false);
  }
  closeOnDelete_ = file;
  FRETURN(true);
}
#undef __FUNC__

#define __FUNC__ "bool SpindleFile::open( const char* filename, const char* openFlags )"
bool SpindleFile::open( const char* filename, const char* openFlags ) {
  FENTER;
  char buffer[128];
  int len = strlen(filename);  // len does not include null char
 
  if (openFlags[0] == 'r') { // read the file
      // check for pipes
      if (strchr(filename,'|') != NULL) {
	strcpy(buffer, filename);  // make a local copy
	while( (buffer[len-1] == ' ') || (buffer[len-1] == '|')) { // remove trailing pipe and spaces.
	  len--;
	  buffer[len] = '\0';
	}
	if ((fp_ = popen( buffer, openFlags )) == NULL) { 
	  cerr << "Error: Cannot open pipe \'" << filename << "\'" <<  endl;
	  FRETURN(false);
	}
	closeOnDelete_ = pipe;
      } 
#ifdef DO_UNCOMPRESS
      else if ( (filename[len-2] == '.') && (filename[len-1]=='Z') ) {
	// open uncompress pipe
	sprintf( buffer, DO_UNCOMPRESS, filename);
	if ((fp_ = popen( buffer, openFlags )) == NULL) { 
	  cerr << "Error: Cannot open pipe \'" << buffer << "\'" <<  endl;
	  FRETURN(false);
	}
	closeOnDelete_ = pipe;
      } 
#endif
#ifdef DO_GUNZIP 
      else if ( (filename[len-3] == '.') && (filename[len-2]=='g') && (filename[len-1]=='z') ) {
	// open gunzip pipe
	sprintf( buffer, DO_GUNZIP, filename);
	if ((fp_ = popen( buffer, openFlags )) == NULL) { 
	  cerr << "Error: Cannot open pipe \'" << buffer << "\'" <<  endl;
	  FRETURN(false);
	}
	closeOnDelete_ = pipe;
      }
#endif
#ifdef DO_PKUNZIP
      else if ( (filename[len-4] == '.') && (filename[len-3]=='z') 
		&&  (filename[len-2]=='i') && (filename[len-1]=='p') ) {
	// open zip pipe
	sprintf( buffer, DO_PKUNZIP, filename);
	if ((fp_ = popen( buffer, openFlags )) == NULL) { 
	  cerr << "Error: Cannot open pipe \'" << buffer << "\'" <<  endl;
	  FRETURN(false);
	}
      } 
#endif
      else {
	// open regular file
	if ((fp_ = fopen( filename, openFlags )) == NULL) { 
	  cerr << "Error: Cannot open file \'" << filename << "\'" << endl;
	  FRETURN(false);
	}
	closeOnDelete_ = file;
      }  
  } else if (openFlags[0]=='w') { // write the file
    // check for pipes
    if (strchr(filename,'|') != NULL) {
      const char * temp = filename;
      while (( *temp == ' ') || (*temp  == '|')) { // skip leading pipe and spaces.
	temp++;
      }
      if ((fp_ = popen( temp, openFlags )) == NULL) { 
	cerr << "Error: Cannot open pipe \'" << filename << "\'" <<  endl;
	FRETURN(false);
      }
      closeOnDelete_ = pipe;
    }
#ifdef DO_COMPRESS
    else if ( (filename[len-2] == '.') && (filename[len-1]=='Z') ) {
      // open uncompress pipe
      sprintf( buffer, DO_COMPRESS, filename);
      if ((fp_ = popen( buffer, openFlags )) == NULL) { 
	cerr << "Error: Cannot open pipe \'" << buffer << "\'" <<  endl;
	FRETURN(false);
      }
      closeOnDelete_ = pipe;
    }
#endif
#ifdef DO_GZIP
    else if ( (filename[len-3] == '.') && (filename[len-2]=='g') && (filename[len-1]=='z') ) {
      // open gunzip pipe
      sprintf( buffer, DO_GZIP, filename);
      if ((fp_ = popen( buffer, openFlags )) == NULL) { 
	cerr << "Error: Cannot open pipe \'" << buffer << "\'" <<  endl;
	FRETURN(false);
      }
      closeOnDelete_ = pipe;
    }
#endif
#ifdef DO_PKZIP
    else if ( (filename[len-4] == '.') && (filename[len-3]=='z') &&  (filename[len-2]=='i') 
	      && (filename[len-1]=='p') ) {
      // open zip pipe
      // sprintf( buffer, ZIP, filename);
      // if ((fp_ = popen( buffer, openFlags )) == NULL) { 
      //    cerr << "Error: Cannot open pipe \'" << buffer << "\'" <<  endl;
      //  FRETURN(false);
      // }
      cerr << "Error: Cannot open pipe \'" << filename << "\'" <<  endl;
      cerr << "       PKZIP can't create zipfiles from a pipe!" << endl;
      FRETURN(false);
    }
#endif
    else {
      // open regular file
      if ((fp_ = fopen( filename, openFlags )) == NULL) { 
	cerr << "Error: Cannot open file \'" << filename << "\'" << endl;
	FRETURN(false);
      }
      closeOnDelete_ = file;
    }  
  } else { 
    cerr << "Error: illegal file flags \"" << openFlags << "\"" << endl;
    FRETURN(false);
  }
  FRETURN(true);
}
#undef __FUNC__


#define __FUNC__ "SpindleFile::~SpindleFile()"
SpindleFile::~SpindleFile() {
  FENTER;
  if ( (fp_ != 0) && (closeOnDelete_ != dont) ) {
    FCALL close();
  }
  decrementInstanceCount( SpindleFile::MetaData );
  FEXIT;
}
#undef __FUNC__

#define __FUNC__ "void SpindleFile::rename(const char* oldFilename, const char* newFilename )"
void
SpindleFile::rename(const char* oldFilename, const char* newFilename ) {
  FENTER;
  rename( oldFilename, newFilename );
  FEXIT;
}
#undef __FUNC__

#define __FUNC__ "void SpindleFile::remove(const char* filename )"
void
SpindleFile::remove(const char* filename) {
  FENTER;
  remove( filename );
  FEXIT;
}
#undef __FUNC__


#define __FUNC__ "void SpindleFile::flush()"
void
SpindleFile::flush() {
FENTER;
  ASSERT( fp_ != 0, "This file is invalid");
  fflush( fp_ );
FEXIT;
}
#undef __FUNC__


#define __FUNC__ "void SpindleFile::close()"
void 
SpindleFile::close() {
  FENTER;
  if ( fp_ != 0 ) {
    FCALL flush();
    if ( closeOnDelete_ == file ) {
      fclose( fp_ );
    } else if ( closeOnDelete_ == pipe ) {
      pclose( fp_ );
    }
  }
  fp_= 0;
  FEXIT;
}
#undef __FUNC__

#define __FUNC__ "long SpindleFile::seek( long offset, size_t from )"
bool
SpindleFile::seek( long offset, SpindleFile::SeekOption from ) {
  FENTER;
  int err = -1;
  if ( fp_ != 0 ) {
    switch ( from ) {
    case FILE_START:
      err = fseek( fp_, offset, SEEK_SET );
      break;
    case FILE_CUR:
      err = fseek( fp_, offset, SEEK_CUR );
      break;
    case FILE_END:
      err = fseek( fp_, offset, SEEK_END );
      break;
    default:
      WARN_IF( true, "Unrecognized seek option %d.", from );
    }
  }
  FRETURN( err != 0 );
}
#undef __FUNC__

#define __FUNC__ "void spindleFile::rewind()"
void
SpindleFile::rewind() {
  FENTER;
  if ( fp_ != 0 ) {
    ::rewind( fp_ );
  }
  FEXIT;
}
#undef __FUNC__

bool 
SpindleFile::isEOF() { 
  return ( feof(fp_) != 0 );
}

bool 
SpindleFile::isError() {
  return ( ferror(fp_) != 0);
}

#define __FUNC__ "size_t SpindleFile::read( void* buffer, size_t nBytes, size_t nObjs  )"
size_t 
SpindleFile::read( void* buffer, size_t nBytes, size_t nObjs  ) {
  FENTER;
  ASSERT( fp_ != NULL, "This file is invalid." );  
  FRETURN( fread( buffer, nBytes, nObjs, fp_ ) );
}
#undef __FUNC__

#define __FUNC__ "size_t SpindleFile::write( const void* buffer, size_t nBytes, size_t nObjs  )"
size_t 
SpindleFile::write( const void* buffer, size_t nBytes, size_t nObjs  ) {
  FENTER;
  ASSERT( fp_ != NULL, "This file is invalid." );
  FRETURN( fwrite( buffer, nBytes, nObjs, fp_ ) );
}
#undef __FUNC__


