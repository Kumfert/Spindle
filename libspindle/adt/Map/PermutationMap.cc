//
// PermutationMap.cc
//
//  $Id: PermutationMap.cc,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is povided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  The PermutationMap class is meant to abstract away the concept of
//  a permutation vector which may be of old2New or new2old varieties.
//
//  A PermutationMap is not in a valid state until finalize() is called
//  to insure that both old2New and new2Old are consistent with each
//  other.
//

#include "spindle/PermutationMap.h"

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

#ifdef __FUNC__
#undef __FUNC__
#endif

PermutationMap::PermutationMap() { 
  incrementInstanceCount( PermutationMap::MetaData );
  currentState = EMPTY;
}

void
PermutationMap::validate() {
  if ( currentState == VALID ) { return; }
  int *src; 
  int *dest;
  if ( changedMostRecent == NEITHER ) {
    src = old2New.begin();
    dest = new2Old.begin();
    if ((*src >= 0)&&(*dest <= 0)) { 
      changedMostRecent = OLD2NEW; 
    } else if ((*src <= 0)&&(*dest >= 0)) { 
      changedMostRecent = NEW2OLD; 
      src = new2Old.begin();
      dest = old2New.begin();
    }
  } else if ( changedMostRecent == OLD2NEW ) { 
    src = old2New.begin();
    new2Old.init( 0 );
    dest = new2Old.begin();
  } else { // if ( changedMostRecent == NEW2OLD ) {
    src = new2Old.begin();
    old2New.init( 0 );
    dest = old2New.begin();
  }
  {for (int i=0; i<sz; i++) { 
    int perm_i = src[i];
    if ( perm_i == -1 ) { continue; }
    if ( ( perm_i < 0) ||  ( perm_i >= sz ) ) { 
      currentState = INVALID;
      return; 
    }
    dest[perm_i]++;
  }}
  // ... now make sure each vertex only occurred once
  int * p = dest;
  { for (int i=0; i<sz; i++) { 
    if ( ( *p != 1 ) && (*p != 0) ) { 
      currentState = INVALID;
      return;
    }
  }}

  // ... finally copy the reverse permutation 
  {for (int i=0; i<sz; i++) { 
    dest[ src[ i ] ] = i; 
  }}
  currentState = VALID;
  changedMostRecent = NEITHER;
  return;
}

bool
PermutationMap::reset() {
  old2New.init( -1 );
  new2Old.init( -1 );
  currentState = EMPTY;
  changedMostRecent = NEITHER;
  return true;
}


const SharedArray<int>& 
PermutationMap::getOld2New() const {
  return old2New;
}

SharedArray<int>& 
PermutationMap::getOld2New() {
  if ( (!isValid() ) && (changedMostRecent == NEW2OLD) ) {
    validate();
    assert( isValid() );
  }
  currentState = UNKNOWN;
  changedMostRecent = OLD2NEW;
  return old2New;
}

const SharedArray<int>& 
PermutationMap::getNew2Old() const {
  return new2Old;
}

SharedArray<int>& 
PermutationMap::getNew2Old() {
  if ((!isValid()) && (changedMostRecent == OLD2NEW) ) {
    validate();
    assert( isValid() );
  }
  currentState = UNKNOWN;
  changedMostRecent = NEW2OLD;
  return new2Old;
}

void 
PermutationMap::resize( const int n ) { 
  currentState = EMPTY;
  changedMostRecent = NEITHER;
  sz = n;
  old2New.resize(n);
  new2Old.resize(n);
}

void
PermutationMap::dump( FILE * fp ) const {
  if ( isValid() ) {
    fprintf(fp,"PermutationMap:");
    const int* p = getNew2Old().lend();
    for( int i=0; i<size(); i++) {
      fprintf(fp,"%d ", p[i] );
    }
  }
}

void
PermutationMap::prettyPrint( FILE * fp, bool new2old_perm, bool printLength, 
			     bool onesOffset ) const { 

  if ( !isValid() ) {
    return ;
  }
  int offset = (onesOffset) ? 1 : 0 ;
  
  if ( printLength == true ) {
    fprintf( fp,"%d\n", size() );
  }
  
  const int * p;
  if ( new2old_perm ) { 
    p = new2Old.lend();
  } else { 
    p = old2New.lend();
  }

  for(const int* cur = p, *stop=p + size(); cur <stop; ++cur ) {
    fprintf( fp, "%d\n", *cur + offset);
  }
} 

void
PermutationMap::storeObject( SpindleArchive& ar ) const {
  if ( isValid() ) {
    const int *p = getNew2Old().lend();
    size_t my_size = size();
    pair< const int*, size_t> my_pair( p,my_size );
    ar << my_pair;
  } else {
    ar << -1;
  }
}
#define __FUNC__ "PermutationMap::loadObject( SpindleArchive& ar )"
void
PermutationMap::loadObject( SpindleArchive& ar ) {
  size_t new_size = ar.peekLength();
  if ( (int)new_size == -1 ) { 
    int junk;
    ar >> junk;
    currentState = INVALID;
  } else {
    resize(new_size);
    int* p = getNew2Old().begin();
    pair<int*, size_t> new_pair( p, new_size );
    
    ar >> new_pair;
    
    WARN_IF( new_pair.second != new_size, "Warning, expected size=%d, but extracted only %d.", 
	     new_pair.second, new_size );
    changedMostRecent = NEW2OLD;
    validate();
  }
}
#undef __FUNC__

SPINDLE_IMPLEMENT_PERSISTANT( PermutationMap, SpindlePersistant )


