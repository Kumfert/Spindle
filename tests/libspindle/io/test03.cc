//
// src/io/drivers/tests/test03.cc -- tests persistance of PermutationMap class
//
//  $Id: test03.cc,v 1.2 2000/02/18 01:32:08 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//
//
//

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifndef SPINDLE_PERMUTATION_H_
#include "spindle/PermutationMap.h"
#endif

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifndef SPINDLE_FILE_H_
#include "spindle/SpindleFile.h"
#endif

#ifdef __FUNC__
#undef __FUNC__
#endif

#define __FUNC__ "main()"
#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

main() {  
  SpindleSystem::registry().setDumpOnDestroy( stdout );
  SpindleSystem::setErrorFile( stdout );
  SpindleSystem::setWarnFile( stdout );

  int perm_array[] = { 3, 4, 2, 0, 1};
  PermutationMap perm(5);
  perm.getNew2Old().take( perm_array, 5 );
  perm.validate();

  {
    SpindleFile outputFile( "temp.out", "w");
    SpindleArchive archive( &outputFile, SpindleArchive::store | SpindleArchive::noByteSwap );
    archive << &perm;
    archive.close();
  }

  SpindleBaseClass* object=0;
  {
    SpindleFile inputFile( "temp.out", "r");
    SpindleArchive  archive2( &inputFile, SpindleArchive::load | SpindleArchive::noByteSwap );
    archive2 >> object;
  }
    
  if ( object == 0 ) { 
    WARNING("Did not retrieve object.");
  } else {
    PermutationMap* perm2 =  PermutationMap::dynamicCast( object );
    if ( perm2 != 0 ) {
      perm2->validate();
      perm2->dump( stdout );
      fprintf( stdout, "\n");
    } else { 
    WARNING("Dynamic Cast failed!");
    }
    delete perm2;
  }
}
