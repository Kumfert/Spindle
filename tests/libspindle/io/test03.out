Permutation:3 4 2 0 1 
Destroying Registry... PersistanceRegistery::dump():  4 items
             SpindleBaseClass (8) 4/4---> 
                  SpindleFile (20) 2/2---> SpindleBaseClass
            SpindlePersistant (12) 2/2---> SpindleBaseClass
                Permutation (44) 2/2---> SpindlePersistant
